### PR TITLE
feat(agents): Antigravity CLI (agy) as a supported provider

### LINKED ISSUE #1381: Feature: Antigravity CLI support
### Summary

Support Antigravity CLI as agent provider

### Problem or opportunity

Ability to orchestrate agent activity with Antigravity CLI.

### Desired outcome

Antigravity CLI is a supported agent provider

### Proposed solution

_No response_

### Additional context

I'm aware of the previous efforts to support Gemini (e.g. #435, #601) but give Google has moved now to Antigravity CLI I wanted to see if there's more folks interested in its support.

### Maestro version

_No response_

### Operating system

None

### Install source

None

### Product area

None

### Agent/provider involved

None

### Is SSH remote execution part of this request?

None

### PR DESCRIPTION
**Pull request by pedramamini**
Closes #1381

Adds Google's Antigravity CLI (`agy`) as a first-class agent provider. This also resolves the dangling `superseded by Antigravity` note that has sat on the hidden `gemini-cli` definition since May.

## What's wired

The full [AGENT_SUPPORT.md](https://github.com/RunMaestro/Maestro/blob/main/AGENT_SUPPORT.md) checklist: agent id, definition, capabilities, display name, beta flag, context window, binary path probing (macOS/Linux `~/.local/bin/agy`, Windows `%LOCALAPPDATA%\agy\bin\agy.exe`), icon, and the new-instance agent list.

| Aspect | Value |
| --- | --- |
| Binary | `agy` |
| Headless | `-p` |
| JSON output | `--output-format stream-json` |
| Resume | `--conversation <id>` |
| Session id field | `conversation_id` |
| Model / effort | `--model <slug>` / `--effort low\|medium\|high` |
| Auto-approve | `--dangerously-skip-permissions` |

## Why a bespoke parser

Antigravity doesn't follow the Claude-family `{"type":"assistant",...}` shape, so it can't be a `ClaudeOutputParser` subclass the way Qwen is. It discriminates on an `event` key and nests the payload under a property of the same name:

```json
{"event":"init","init":{"cwd":"...","model":"...","tools":[...]}}
{"event":"step_update","step_update":{"conversation_id":"...","state":"ACTIVE","step_type":"agent_response","text_delta":"..."}}
{"event":"result","result":{"conversation_id":"...","status":"...","response":"...","usage":{...}}}
```

`AntigravityOutputParser` maps `step_update` text deltas to partial text, tool steps to `tool_use`, and normalizes the snake_case token metrics (`input_tokens` / `output_tokens` / `thinking_tokens` / `cache_read_tokens`). A `result` envelope carrying `error` is reclassified as an error event: `error` is documented as present *only* on failure, which is a firmer signal than the free-form `status` string, so nothing is inferred from status values the docs never enumerate.

## Two headless footguns handled in the definition

- **Batch runs pass `--dangerously-skip-permissions`.** Without it the CLI soft-denies shell commands, so a run stalls on its first terminal tool call rather than failing loudly. `buildAgentArgs` already drops `batchModeArgs` in read-only mode, so read-only still gets `--sandbox` and no permission skip.
- **`--print-timeout` defaults to 5m**, which silently cuts off longer agent runs. Exposed as a "Headless Timeout" config option defaulting to 30m.

## Deliberately conservative capabilities

Anything the docs don't state outright is left `false` rather than guessed at:

- `supportsSessionStorage: false` - the on-disk conversation format is undocumented.
- `supportsImageInput: false` - no documented attachment flag.
- `supportsReadOnlyMode: false` - headless mode auto-allows reading **and writing** files inside the active workspace; `--sandbox` only restricts terminal commands. There is no flag that makes the agent truly read-only.
- `supportsSlashCommands: false` - slash commands are a TUI-only affordance.
- `supportsThinkingDisplay: false` - `thinking_tokens` are counted, but no reasoning-text step type is documented.

## Honest caveat

**This is built against the published headless contract, not a captured live run.** I don't have `agy` installed to verify byte-for-byte against real output, which is why it ships as beta (`BETA_AGENTS`). The event shape, flags, and usage field names all come from [antigravity.google/docs/cli/headless](https://antigravity.google/docs/cli/headless). If anything drifts, the parser and the definition are the two places to touch, and the schema is documented inline in both.

Auth note for testers: headless runs reuse cached credentials, so you need to sign in once from an interactive `agy` session on the machine first.

## Testing

- New: `src/__tests__/main/parsers/antigravity-output-parser.test.ts` (13 tests) covering each event type, usage normalization, failed-result reclassification, error-pattern classification, non-JSON passthrough, and exit-code handling.
- New: 5 cases in `agent-args.test.ts` asserting the composed command line off the real definition, including that read-only drops the permission skip and that the 30m timeout applies with no user config.
- Updated hardcoded agent/parser counts in `detector.test.ts` and `parsers/index.test.ts`.
- `npm run lint`, `npm run lint:eslint`, and the full suite (1170 files, 32,952 tests) pass locally on macOS. Both CI matrix legs still need to confirm.

## Docs

`AGENT_SUPPORT.md` gains a full Antigravity reference section (flags, event schema, limitations, sources); the stale "Gemini CLI Planned" section now points at it. `CLAUDE.md`'s supported-agents table gains the row.

## Summary by CodeRabbit

* **New Features**
  * Added beta support for Antigravity CLI, including detection, conversation resumption, streaming output, model and reasoning controls, and usage metrics.
  * Added automatic discovery of Antigravity installations across supported platforms.
  * Centralized provider selection across the New Agent modal, wizard, and group-chat moderator controls.
  * Added a horizontally scrollable provider strip with navigation controls and new provider logos.
* **Bug Fixes**
  * Improved handling of authentication, quota, timeout, network, tool, and CLI errors.
  * Improved reporting of streaming command failures without duplicate or misleading results.
* **Documentation**
  * Added Antigravity setup, authentication, capabilities, limitations, and usage guidance.

### CODE REVIEW

**Comment by pedramamini** [comment]
Pushed f59859b addressing the automated review. Both bots landed on real problems.

**`isResultMessage` could surface a failure as the agent's answer** (Greptile). It also returned true when the raw payload was `event: 'result'` - which is still true for a *failed* envelope after `parseResult` reclassifies it to `type: 'error'`. `StdoutHandler` returns early on error events before emitting a result, so the streaming path was safe, but `ExitHandler`'s end-of-stream flush (the path for a last line with no trailing newline) keys off `isResultMessage` alone and emits `event.text` to the user. Narrowed to `event.type === 'result'`, which is what `ClaudeOutputParser` does and what `QwenOutputParser` relies on. My original test asserted the buggy behavior; it now asserts `false` and says why.

**A conversation could be stranded without a resumable id** (CodeRabbit). The documented `init` payload has no `conversation_id` - it first appears on `step_update` - so a run dying immediately after init left nothing to resume from. `extractSessionId` now reads a top-level `conversation_id` defensively, with a test covering it. Costs nothing if the CLI never emits one.

**Tab naming exposure** (Greptile): confirmed real, but not fixable here. Tab naming spawns read-only and relies on `noToolsArgs` to keep a task-like first message from becoming a real agentic run; Antigravity has no tool-disabling flag to put there. Worth noting `noToolsArgs` is Claude-only today, so every non-Claude agent shares this - Antigravity is just the one that *also* lacks CLI-enforced read-only, so it can reach workspace writes where Codex can't. Documented in both the definition and `AGENT_SUPPORT.md` with a note to wire it up if such a flag ships.

Full suite green locally (1170 files, 32,953 tests), lint and eslint clean.

**Comment by martingalloar** [comment]
I've been using this branch with the Antigravity CLI for some days. Functionally all good, there's some missing items that were noted there (e.g. no thinking output, no model/reasoning dropdown selection) but so far so good.

**Comment by pedramamini** [comment]
## Retargeted to `rc`, and completed the registration

Rebased onto `rc` (was `main`) and resolved the drift: `AGENT_LOGIN_COMMANDS` and the two capability fields `rc` has added since this branch was cut (`supportsPromptViaStdin`, `supportsAdditionalDirectories`) both needed Antigravity entries or the branch would not compile on `rc`.

### The PR was not complete

Antigravity was defined, capable, and detected, but only reachable in **one** of the three places a user picks a provider:

| Surface | Before | After |
| --- | --- | --- |
| New Agent modal | shown | shown |
| New Agent Wizard tile strip | **missing** | shown |
| Group Chat moderator dropdown | **missing** | shown |

It was not an Antigravity-specific miss. `AGENT_TILES` (wizard) and `SUPPORTED_AGENTS` (New Agent modal) were two hand-written lists of agent ids with nothing forcing them to agree, and the moderator dropdown reads the first one. **Grok and Qwen3 Coder had shipped that way for months**: selectable in the New Agent modal, absent from the wizard, un-pickable as a moderator.

### One registry, three surfaces

`AGENT_PICKER_META` in `src/shared/agentMetadata.ts` is now the single source. It is a `Record<AgentId, AgentPickerMeta | null>`, so a new agent id **does not compile** until it is either given picker metadata or explicitly set to `null` (`terminal`, `gemini-cli`). Key order is picker order.

- `AGENT_TILES` is derived from it, `SUPPORTED_AGENTS` re-exports it, the moderator dropdown filters those same tiles by what detection found.
- Antigravity, Grok, and Qwen3 Coder get real tile marks. A test fails if a pickable provider would draw the blank fallback ring.
- Brand colors run through `readableTextOn()`, so Copilot's near-black `#24292F` stops vanishing into a dark tile.

Note: the `supportsGroupChatModeration` capability flag is advisory - nothing reads it, and the dropdown offers any installed provider. Left as-is, since gating on it would cut the list to two.

### Wizard: single scrolling row

Eleven providers wrapped to a third row and pushed Continue below the fold. The grid is now one horizontally scrolling row of slightly larger tiles, with edge fades, arrow buttons that only appear when they can actually move, wheel-to-horizontal, and left/right arrow-key stepping. New shared hook: `useHorizontalScroll` (`hooks/ui/`).

Removed as now-unreachable: `GRID_COLS`, `GRID_ROWS`, `LAST_ROW_START_INDEX`, `LAST_ROW_COL_START_CLASS`, `getAgentTileColSpanClass`, and the `colSpanClass` tile prop.

### Docs

The "how do I register a provider" gap is closed in four places: `AGENT_SUPPORT.md` gains **Step 2.6** and three new required-steps entries, plus `CLAUDE-AGENTS.md`, `docs/agent-guides/AGENT-INFRA.md`, and `CLAUDE.md`'s canonical-utilities list.

### Validation

`tsc` clean on all four configs, ESLint clean, and the full suite green via the pre-push hook: **1560 files, 37,571 tests passed**. Both CI matrix legs still need to confirm.

File: src/main/agents/definitions.ts [hunk 1]
```diff
@@ -346,6 +346,73 @@ export const AGENT_DEFINITIONS: AgentDefinition[] = [
...
+		jsonOutputArgs: ['--output-format', 'stream-json'],
+		// `agy -p "..." --conversation <id>` resumes a specific conversation. The id
+		// comes back as `conversation_id` on every stream-json event.
+		resumeArgs: (sessionId: string) => ['--conversation', sessionId],
+		// Best-effort only: `--sandbox` restricts terminal commands, but reading AND
+		// writing files inside the active workspace stays auto-allowed in headless
+		// mode. There is no CLI flag that makes the agent truly read-only, hence
+		// readOnlyCliEnforced: false and supportsReadOnlyMode: false.
+		readOnlyArgs: ['--sandbox'],
+		readOnlyCliEnforced: false,
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/main/agents/definitions.ts:401]: Half of this is real, half is defused by the caller. Taking them separately.

**"Times out without naming the tab"** - it cannot run long. `handleTabNamingRequest` kills the spawn at `TAB_NAMING_TIMEOUT_MS = 120s` (`src/main/ipc/handlers/tabNaming.ts:76`, `:452`), so the 30m `--print-timeout` never governs. A soft-denied tool call costs a failed naming attempt, not a runaway process. Same outcome the path already has for any agent whose first message reads like a task.

**"Can modify project files"** - correct, and deliberately documented at the definition (`definitions.ts:394-399`). Worth being precise about what is new here, though: this is a property of the naming path for **any** agent with `readOnlyCliEnforced: false` and no `noToolsArgs`, and Antigravity is the second such agent, not the first. `qwen3-coder` has shipped with exactly that pair since it was added (`definitions.ts:336-337`: `readOnlyArgs: ['-y']`, `readOnlyCliEnforced: false`, no `noToolsArgs`). `buildAgentArgs` already logs a warning on every such spawn (`agent-args.ts:240`).

I considered the obvious guard - have the naming path refuse to spawn when an agent can neither disable tools nor enforce read-only - and rejected it for this PR:

- It is not an Antigravity fix. It changes `qwen3-coder` too, and silently removes tab naming for those users to close a hole nobody has reported hitting.
- It belongs with the naming path, not with a provider definition, so it should be its own PR with its own test coverage and its own call on whether losing naming beats the risk.

The honest state is what the code already says: `--sandbox` restricts terminal commands only, there is no flag that makes `agy` truly read-only, hence `supportsReadOnlyMode: false` and `readOnlyCliEnforced: false`, and `noToolsArgs` gets wired the moment such a flag ships.

The other P1 on this PR (**error envelopes become successful results**) was a real bug and is fixed in `676c1ffde`: `isResultMessage` now gates on `event.type === 'result'` rather than the raw envelope, so a failed `result` reclassified to `type: 'error'` no longer reaches the user through ExitHandler's end-of-stream flush.

**Comment by pedramamini** [comment]
@coderabbitai full review

The branch has been rebased from `main` onto `rc` and has grown well past the original Antigravity scope. Please review the whole diff fresh rather than incrementally. Areas I would most like a second pair of eyes on:

1. `AGENT_PICKER_META` in `src/shared/agentMetadata.ts` - a `Record<AgentId, AgentPickerMeta | null>` that is now the single source for all three provider pickers. Is the `null`-means-withheld encoding sound, and does anything still read a hand-written agent list?
2. `useHorizontalScroll` (`src/renderer/hooks/ui/useHorizontalScroll.ts`) - listener lifecycle, the non-passive `wheel` handler, and the jsdom no-op path.
3. The wizard strip's keyboard model after dropping the 2-D grid: `getNextAgentTileIndex` now only steps left/right.

File: src/main/process-manager/handlers/ExitHandler.ts [hunk 2]
```diff
@@ -137,7 +137,24 @@ export class ExitHandler {
...
+				// A terminal envelope that reports a FAILURE has to leave through the
+				// error path, not the result path. Emitting its text as data would
+				// render a provider failure as the agent's answer, and dropping it
+				// silently is worse still: `detectErrorFromExit` below returns null on
+				// exit code 0, so a CLI that reports the failure in-band and then exits
+				// clean would settle the turn with no answer and no error at all - the
+				// tab just stops, and no retry or recovery handling ever fires.
+				if (event?.type === 'error' && !managedProcess.errorEmitted) {
+					const agentError = outputParser.detectErrorFromParsed((event.raw as unknown) ?? event);
+					if (agentError) {
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/main/process-manager/handlers/ExitHandler.ts]: Agreed, and fixed in `bfc808f52`.

You are right about the direction of the failure, and it is worth naming how bad it was: the widened catch would have produced *the exact bug the commit set out to close*. A throwing classifier would be swallowed, control would fall into the `catch`, and the failed envelope's raw JSON would be emitted through `emitDataBuffered` - the provider failure rendered as the agent's answer, just in JSON rather than prose.

The parse now sits alone in the `try`. Classification and dispatch run after it and are free to bubble to Sentry:

```ts
let event: ParsedEvent | null = null;
try {
    event = outputParser.parseJsonLine(remainingLine);
} catch {
    this.bufferManager.emitDataBuffered(sessionId, remainingLine, managedProcess);
}

if (event?.type === 'error' && !managedProcess.errorEmitted) { ... }
else if (event && outputParser.isResultMessage(event) && ...) { ... }
```

A malformed last line stays an expected, recoverable condition with a defined fallback, which is the one thing that catch was ever for. Everything past it is unexpected and now crashes honestly.

Two tests added, both verified to fail against the previous shape:

- a throwing `detectErrorFromParsed` propagates out of `handleExit` and the envelope never appears in the `data` stream
- a malformed trailing line still falls back to raw emission, so the original behaviour is pinned

Worth noting the same widening covered the result branch too, so `isResultMessage` or `emitDataBuffered` throwing had the identical leak. That is closed by the same change.

File: CLAUDE-AGENTS.md [hunk 3]
```diff
@@ -136,9 +163,13 @@ To add support for a new agent:
...
-9. (Optional) Add error patterns to `src/main/parsers/error-patterns.ts`
+6. **Register it in the pickers**: add an entry to `AGENT_PICKER_META` in `src/shared/agentMetadata.ts`, or `null` to withhold it. This is what makes the provider appear in the New Agent modal, the wizard tile strip, and the Group Chat moderator dropdown
+7. Add a re-auth command to `AGENT_LOGIN_COMMANDS` in `src/shared/agentMetadata.ts`
+8. Draw the tile logo: a `case` in `AgentSelectionScreen/components/AgentLogo.tsx`, plus a glyph in `src/renderer/constants/agentIcons.ts`
+9. Add install locations to `src/main/agents/path-prober.ts` if the binary does not always land on `$PATH`
+10. Sync `AgentCapabilities` interface in renderer: `useAgentCapabilities.ts`, `types/index.ts`, `global.d.ts`
+11. (If `supportsJsonOutput`) Create output parser in `src/main/parsers/{agent}-output-parser.ts`, register in `src/main/parsers/index.ts`
+12. (If `supportsSessionStorage`) Create session storage extending `BaseSessionStorage` in `src/main/storage/`
+13. (Optional) Add error patterns to `src/main/parsers/error-patterns.ts`
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply CLAUDE-AGENTS.md]: Fixed in `c334a28db`. It now reads "(If the agent has an output parser) ... required, not optional", and names `agent-completeness.test.ts` as the thing that fails CI. You are right that the two docs contradicted each other and that mine was the wrong one.

File: src/__tests__/main/agents/agent-completeness.test.ts [hunk 4]
```diff
@@ -164,6 +169,36 @@ describe('Agent Completeness', () => {
...
+		});
+
+		it('every pickable agent can be re-authenticated from the UI', () => {
+			// A provider offered in a picker whose auth expires needs a way back in;
+			// null is only correct for agents that carry no credentials of their own.
+			for (const agentId of PICKABLE_AGENT_IDS) {
+				const login = getAgentLoginCommand(agentId);
+				if (login === null) continue;
+				expect(login.binary, `Agent "${agentId}" has an empty login binary`).toBeTruthy();
+			}
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/__tests__/main/agents/agent-completeness.test.ts:225]: Fixed in `c334a28db`. The `continue` made the assertion unreachable for exactly the case worth catching - a missing `AGENT_LOGIN_COMMANDS` entry was indistinguishable from a deliberate credentialless one.

Credentialless agents are now a named set that must be `null`, so the claim is explicit and reversible:

```ts
const CREDENTIALLESS_AGENTS = new Set(["hermes", "pi", "omp"]);
```

Every other pickable agent must supply a non-null command, a non-empty `binary`, and a string `args`. The set is asserted in both directions: an agent listed there that later gains a login command fails too, so the list cannot quietly rot.

File: src/__tests__/main/parsers/antigravity-output-parser.test.ts [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+		const event = parser.parseJsonObject({
+			event: 'init',
+			init: {
+				cwd: '/tmp/project',
+				tools: ['read_file', 'run_command'],
+				permission_mode: 'skip',
+				model: 'gemini-3.6-flash-high',
+			},
+		});
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/__tests__/main/parsers/antigravity-output-parser.test.ts:22]: Fixed in `c334a28db` - fixtures now use `SUCCESS`, `ERROR`, and `always-proceed`.

Went further on the point behind it. Since the parser deliberately ignores `status`, renaming the strings alone documents a contract without testing it, so there is a new test that pins the real rule with two envelopes that contradict their own status:

- `{ status: "SUCCESS", error: "RESOURCE_EXHAUSTED: ..." }` must classify as an error
- `{ status: "ERROR", response: "fine" }` must not

That fails if anyone ever "helpfully" starts reading `status`, which the free-form-string caveat in `AGENT_SUPPORT.md` warns against.

File: src/main/parsers/antigravity-output-parser.ts [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+
+		const match = matchErrorPattern(getErrorPatterns(this.agentId), errorText);
+		return {
+			type: match?.type ?? 'unknown',
+			message: match?.message ?? errorText,
+			recoverable: match?.recoverable ?? true,
+			agentId: this.agentId,
+			timestamp: Date.now(),
+			parsedJson: parsed,
+		};
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/main/parsers/antigravity-output-parser.ts:346]: Fixed in `c334a28db`, and the consequence is worse than losing an association: without `conversation_id` a **recoverable** error cannot be retried with `--conversation`, so the retry would silently open a fresh conversation instead of resuming the failed one, and the user would lose their context with no visible failure.

`sessionId` is now carried from `parsed.result.conversation_id`, with a test asserting it survives classification.

File: src/__tests__/renderer/components/Wizard/screens/AgentSelectionScreen/components.test.tsx [hunk 7]
```diff
@@ -53,6 +54,18 @@ describe('AgentSelectionScreen components', () => {
...
 
+	it('draws a real mark for every provider a user can pick', () => {
+		// The fallback is an empty circle. A provider that reaches a picker without
+		// a mark of its own renders as that blank ring, which reads as a bug.
+		for (const agentId of PICKABLE_AGENT_IDS) {
+			const { container, unmount } = render(
+				<AgentLogo agentId={agentId} supported detected theme={mockTheme} />
+			);
+			expect(container.querySelector('svg'), `${agentId} has no logo`).toBeInTheDocument();
+			unmount();
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/__tests__/renderer/components/Wizard/screens/AgentSelectionScreen/components.test.tsx:72]: Fixed in `c334a28db`, though one premise needs correcting: the fallback is a `<div>`, not an SVG -

```tsx
default:
    return <div className={`${LOGO_SIZE_CLASS} rounded-full border-2`} ... />;
```

so `querySelector("svg")` did reject it and the test was catching a missing mark today. The recommendation is still right, for the reason you give one step out: the assertion is coupled to the fallback happening not to be an SVG. Redraw that ring as an `<svg>` some day and the test keeps passing for precisely the providers it exists to catch, silently.

So it now asserts on the fallback itself rather than on the shape of what a real logo happens to be. `AgentLogo` exports `AGENT_LOGO_FALLBACK_TESTID`, every `PICKABLE_AGENT_IDS` entry must render without it, and a companion test renders an unknown provider to prove the marker actually appears - otherwise the coverage test could pass by the marker having been removed.

File: src/renderer/hooks/ui/useHorizontalScroll.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+		// jsdom has no layout engine and no ResizeObserver by default; skip the
+		// observer rather than throwing so component tests render without a polyfill.
+		const observer =
+			typeof ResizeObserver === 'undefined' ? null : new ResizeObserver(() => measure());
+		observer?.observe(element);
+
+		return () => {
+			element.removeEventListener('scroll', measure);
+			element.removeEventListener('wheel', onWheel);
+			observer?.disconnect();
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/renderer/hooks/ui/useHorizontalScroll.ts]: Fixed in `c334a28db`. Both listeners now go through `useEventListener()`, with `passive: false` preserved on `wheel`.

Worth recording why the raw pairs were there, since it is a trap for the next person: `useEventListener` takes the element, not a ref, and a ref filling in does not trigger a re-render - so passing `ref.current` subscribes to `null` on the mount render and never re-subscribes. The hook now mirrors the element into state, so the listeners bind on the render after mount:

```ts
const [element, setElement] = useState<HTMLElement | null>(null);
useEffect(() => { setElement(ref.current); }, [ref, resetKey]);
```

The ResizeObserver stays in its own effect keyed on `element`, since that has no `useEventListener` equivalent.

File: src/renderer/hooks/ui/useHorizontalScroll.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+		// A strip has no vertical overflow of its own, so a wheel gesture over it
+		// would otherwise scroll an ancestor (or nothing at all).
+		const onWheel = (event: WheelEvent) => {
+			const delta = Math.abs(event.deltaX) > Math.abs(event.deltaY) ? event.deltaX : event.deltaY;
+			if (delta === 0) return;
+			const maxScrollLeft = element.scrollWidth - element.clientWidth;
+			if (maxScrollLeft <= 0) return;
+			event.preventDefault();
+			element.scrollLeft = Math.max(0, Math.min(maxScrollLeft, element.scrollLeft + delta));
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/renderer/hooks/ui/useHorizontalScroll.ts]: Good catch, and fixed in `c334a28db`. The ordering was simply wrong: `preventDefault()` fired before the clamp was computed, so at an end stop the gesture was consumed and nothing moved. The concrete symptom is a dead zone - park the pointer over the strip, scroll further left at the left edge, and neither the strip nor the wizard behind it moves.

Now the clamped target is computed first and the event is only claimed when it actually changes the position:

```ts
const target = Math.max(0, Math.min(maxScrollLeft, element.scrollLeft + delta));
if (target === element.scrollLeft) return;
wheel.preventDefault();
element.scrollLeft = target;
```

The existing `maxScrollLeft <= 0` early return already covered the no-overflow case; this covers the at-an-edge case it missed.

File: src/__tests__/main/agents/agent-completeness.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/__tests__/main/agents/agent-completeness.test.ts:225]: @coderabbitai `c334a28dbfe5ee62a03c95da192b36564c3ec4a6` is now on `origin/feat/1381-antigravity-cli-support` and is the PR head - the earlier check raced the push, which was held up by the pre-push suite. Please re-verify.

File: src/__tests__/main/parsers/antigravity-output-parser.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/__tests__/main/parsers/antigravity-output-parser.test.ts:22]: @coderabbitai `c334a28dbfe5ee62a03c95da192b36564c3ec4a6` is now on `origin/feat/1381-antigravity-cli-support` and is the PR head - the earlier check raced the push, which was held up by the pre-push suite. Please re-verify.

File: src/main/parsers/antigravity-output-parser.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/main/parsers/antigravity-output-parser.ts:346]: @coderabbitai `c334a28dbfe5ee62a03c95da192b36564c3ec4a6` is now on `origin/feat/1381-antigravity-cli-support` and is the PR head - the earlier check raced the push, which was held up by the pre-push suite. Please re-verify.

File: src/renderer/hooks/ui/useHorizontalScroll.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/renderer/hooks/ui/useHorizontalScroll.ts]: @coderabbitai `c334a28dbfe5ee62a03c95da192b36564c3ec4a6` is now on `origin/feat/1381-antigravity-cli-support` and is the PR head - the earlier check raced the push, which was held up by the pre-push suite. Please re-verify.

File: src/main/process-manager/handlers/ExitHandler.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/main/process-manager/handlers/ExitHandler.ts]: noop

**Comment by pedramamini** [comment]
@greptileai review

Your **buffered failures are silently dropped** finding is fixed across `ba99b4583` + `bfc808f52`. Please confirm, and take a fresh pass over the rest.

You were right that `676c1ffde` only closed half the hole. `isResultMessage` stopped a failed envelope being read as the answer, but gave it nowhere else to go: ExitHandler's end-of-stream flush only ever emitted result messages, and `detectErrorFromExit` returns `null` on exit code 0, so an in-band failure followed by a clean exit settled the turn with no answer and no error.

The flush now routes an error event through `detectErrorFromParsed` and emits `agent-error`, mirroring StdoutHandler (`errorEmitted` guard and SSH remote tag included). It lives in ExitHandler, not the Antigravity parser, because every stream-json parser reaches the same flush.

`bfc808f52` then narrowed the surrounding `catch` to the JSON parse alone - it had grown to wrap classification, which would have re-created the original bug by emitting the failed envelope's raw JSON as data.

Four tests, each verified to fail against the prior shape: error leaves as an error and never as data; an error already reported from stdout is not double-reported; a throwing classifier propagates rather than leaking the envelope; a malformed trailing line still falls back to raw emission.

Also worth a look: CodeRabbit's seven findings from its full-diff pass are all addressed in `c334a28db` (wheel end-stop `preventDefault`, `useEventListener` migration, `conversation_id` on structured errors, and two tests that could pass vacuously).

File: src/main/process-manager/handlers/ExitHandler.ts [hunk 10]
```diff
@@ -135,19 +136,42 @@ export class ExitHandler {
...
+			if (event?.type === 'error' && !managedProcess.errorEmitted) {
+				const agentError = outputParser.detectErrorFromParsed((event.raw as unknown) ?? event);
+				if (agentError) {
+					managedProcess.errorEmitted = true;
+					agentError.sessionId = sessionId;
+					if (managedProcess.sshRemoteId) {
+						agentError.sshRemoteId = managedProcess.sshRemoteId;
+					}
+					this.emitter.emit('agent-error', sessionId, agentError);
+				}
```
[the remark this answers is not in the document]
Comment (pedramamini) [reply src/main/process-manager/handlers/ExitHandler.ts:186]: Correct, and fixed in `6a8eb9d0e`. Good catch - this is the third layer of the same envelope and the one with the sharpest consequence.

The flush dispatched the event but never recorded the provider's session id. On the path I had just added, that means a **recoverable** error with no recorded `conversation_id` cannot be retried with `--conversation`, so recovery silently opens a fresh conversation and discards exactly the context the retry existed to continue. Silent, and worse than the failure it was recovering from.

Chasing it down showed the gap was not error-specific: a *successful* trailing result had the identical hole. It just cost resume instead of recovery, so nobody had noticed. Both branches now capture before dispatch, mirroring `emitSessionIdIfNeeded`:

```ts
if (event) {
    const eventSessionId = outputParser.extractSessionId(event);
    if (eventSessionId) {
        managedProcess.agentSessionId = eventSessionId;   // always record
        if (!managedProcess.sessionIdEmitted) {           // emit once
            managedProcess.sessionIdEmitted = true;
            this.emitter.emit('session-id', sessionId, eventSessionId);
        }
    }
}
```

Three tests, the first two verified to fail against the previous shape: the id is captured from a failed envelope, and from a successful one, and an id already reported mid-stream is still recorded on the process without emitting a duplicate event.

Tally for this PR, since all three of your findings came off one thread: `676c1ffde` stopped a failure being read as the answer, `ba99b4583` gave that failure somewhere to go, `bfc808f52` stopped the catch re-creating the first bug, and `6a8eb9d0e` keeps the conversation resumable through it. Each one only became visible once the previous was fixed.
